http-server/src/server.rs

```diff
@@ -578,80 +640,28 @@ pub struct Server<L = ()> {
 	tokio_runtime: Option<tokio::runtime::Handle>,
 	logger: L,
 	health_api: Option<HealthApi>,
+	service_builder: tower::ServiceBuilder<B>,
 }
 
-impl<L: Logger> Server<L> {
+impl<B, U, L: Logger> Server<B, L>
+where
+	B: Layer<TowerService<L>> + Send + 'static,
```

Comment:
:smile: 

hopefully, we don't have to bother with these trait bounds anymore

Reply:
Do you mean there is some way to ismplify them?

Nit: If not, perhaps they should just be on the `start` method and not on the whole impl block (not a big deal but just that `local_addr()` doesn't need them)

Reply:
I'm almost wondering whether it's worth doing something like this (assuming it's possible):

```rust
trait HttpServiceLayer {} // name up for debate; just a placeholder!

impl <B, U> HttpServiceLayer for B where
   // ... complicated trait bounds copied here
```

And then we'd just have to do `impl <B: HttpServiceLayer, L: Logger> Server<B, L> { .... }` here.

Also then we might use the same bounds on `set_middleware` just to catch any issues with bad user submitted middleware sooner (optional though; better or not?).

Reply:
I mean that the these trait bounds are super hard to get right and takes a lot of time to figure out what the compilation comes from :P

Reply:
Oooh gotcha!

Reply:
(I had a go at splitting the trait bounds into a separate trait, but it was rather messy so I'm happy with it as is!)